oidcserverv2/server.go

```diff
@@ -199,6 +230,107 @@ func (s *Server) Authenticate(ctx context.Context, authID string, ident oidcserv
 		return "", err
 	}
 
+	if s.consentAll ||
+		(s.consentOffline && strContains(sess.LoginRequest.Scopes, "offline_access")) {
+		return fmt.Sprintf("/consent?authid=%s", authID), nil
+	}
+	return s.finishAuthenticate(ctx, authID)
+}
+
+func (s *Server) consent(w http.ResponseWriter, req *http.Request) {
+	if req.Method == "POST" {
+		s.finalizeConsent(w, req)
+		return
+	}
+	s.renderConsent(w, req)
+}
+
+func (s *Server) renderConsent(w http.ResponseWriter, req *http.Request) {
+	l := s.logger.WithField("fn", "renderConsent")
+	ctx := req.Context()
+
+	authID := req.URL.Query().Get("authid")
+	if authID == "" {
+		http.Error(w, "Missing auth ID", http.StatusBadRequest)
+		return
+	}
+
+	sess := &storagev2beta1.Session{}
+	if _, err := s.sessionMgr.storage.Get(ctx, sessKeyspace, authID, sess); err != nil {
+		l.WithError(err).Error("failed to get session")
+		http.Error(w, "Internal error", http.StatusInternalServerError)
+		return
+	}
+
+	cl, err := s.clients.clients.GetClient(sess.LoginRequest.ClientId)
+	if err != nil {
+		l.WithError(err).WithField("client-id", sess.LoginRequest.ClientId).Error("couldn't find client")
+		http.Error(w, "Internal error", http.StatusInternalServerError)
+		return
+	}
+
+	cname := sess.LoginRequest.ClientId
+	if cl.Name != "" {
+		cname = cl.Name
+	}
+
+	td := &consentData{
+		AuthID:     authID,
+		ClientName: cname,
+		Offline:    strContains(sess.LoginRequest.Scopes, "offline_access"),
```

Comment:
Optional: Consider making `"offline_access"` a constant since it's used multiple times.